org.jacoco.core/src/org/jacoco/core/internal/instr/Companions.java

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2016 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.instr;
+
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import org.jacoco.core.internal.Java9Support;
+import org.jacoco.core.internal.data.CRC64;
+import org.jacoco.core.internal.flow.ClassProbesAdapter;
+import org.jacoco.core.runtime.RuntimeData;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * Instrumentation with creation of "companion" classes in "online" mode, which
+ * will hold state of instrumented classes. This allows to avoid addition of
+ * members to classes.
+ */
+public class Companions {
+
+	/**
+	 * {@link ClassLoader#defineClass(String, byte[], int, int)}
+	 */
+	private static final Method DEFINE_CLASS;
+
+	static {
+		try {
+			DEFINE_CLASS = ClassLoader.class.getDeclaredMethod("defineClass",
+					String.class, byte[].class, Integer.TYPE, Integer.TYPE);
+			DEFINE_CLASS.setAccessible(true);
+		} catch (NoSuchMethodException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static final String COMPANION_NAME = InstrSupport.DATAFIELD_NAME;
+
+	private static final String INIT_METHOD_NAME = InstrSupport.INITMETHOD_NAME;
+
+	/**
+	 * https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.11
```

Comment:
This JVMS reference is interesting, but it no longer justifies the current value, so a comment on how this value was chosen would be useful, even if arbitrary or to reduce the class memory size.  Or is it entirely chosen by the maximum method size needed by OfflineInstrumentationCompanionAccessGenerator, or is it arbitrarily lower?  If you're trying to maximize this value, I guess the dynamically generated classes could be larger.